src/Databases/PostgreSQL/DatabasePostgreSQL.cpp

```diff
@@ -0,0 +1,385 @@
+#include <Parsers/queryToString.h>
+#include <Common/escapeForFileName.h>
+#include <Poco/DirectoryIterator.h>
+#include <Poco/File.h>
+#include <Databases/PostgreSQL/FetchFromPostgreSQL.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NOT_IMPLEMENTED;
+    extern const int UNKNOWN_TABLE;
+    extern const int TABLE_IS_DROPPED;
+    extern const int TABLE_ALREADY_EXISTS;
+}
+
+static const auto suffix = ".removed";
+static const auto cleaner_reschedule_ms = 60000;
+
+DatabasePostgreSQL::DatabasePostgreSQL(
+        const Context & context,
+        const String & metadata_path_,
+        const ASTStorage * database_engine_define_,
+        const String & dbname_,
+        const String & postgres_dbname,
+        PGConnectionPtr connection_,
+        const bool cache_tables_)
+    : IDatabase(dbname_)
+    , global_context(context.getGlobalContext())
+    , metadata_path(metadata_path_)
+    , database_engine_define(database_engine_define_->clone())
+    , dbname(postgres_dbname)
+    , connection(std::move(connection_))
+    , cache_tables(cache_tables_)
+{
+    cleaner_task = context.getSchedulePool().createTask("PostgreSQLCleanerTask", [this]{ removeOutdatedTables(); });
+    cleaner_task->deactivate();
+}
+
+
+bool DatabasePostgreSQL::empty() const
+{
+    std::lock_guard<std::mutex> lock(mutex);
+
+    auto tables_list = fetchTablesList();
+
+    for (const auto & table_name : tables_list)
+        if (!detached_or_dropped.count(table_name))
+            return false;
+
+    return true;
+}
+
+
+DatabaseTablesIteratorPtr DatabasePostgreSQL::getTablesIterator(
+        const Context & context, const FilterByNameFunction & /* filter_by_table_name */)
+{
+    std::lock_guard<std::mutex> lock(mutex);
+
+    Tables tables;
+    auto table_names = fetchTablesList();
+
+    for (const auto & table_name : table_names)
+        if (!detached_or_dropped.count(table_name))
+            tables[table_name] = fetchTable(table_name, context, true);
+
+    return std::make_unique<DatabaseTablesSnapshotIterator>(tables, database_name);
+}
+
+
+std::unordered_set<std::string> DatabasePostgreSQL::fetchTablesList() const
+{
+    std::unordered_set<std::string> tables;
+    std::string query = "SELECT tablename FROM pg_catalog.pg_tables "
+        "WHERE schemaname != 'pg_catalog' AND schemaname != 'information_schema'";
+    /// Already connected to the needed database, search will be done there
+    pqxx::read_transaction tx(*connection->conn());
+
+    for (auto table_name : tx.stream<std::string>(query))
+        tables.insert(std::get<0>(table_name));
+
+    return tables;
+}
+
+
+bool DatabasePostgreSQL::checkPostgresTable(const String & table_name) const
+{
+    pqxx::nontransaction tx(*connection->conn());
+    pqxx::result result = tx.exec(fmt::format(
+        "SELECT tablename FROM pg_catalog.pg_tables "
+        "WHERE schemaname != 'pg_catalog' AND schemaname != 'information_schema' AND tablename = '{}'", table_name));
```

Comment:
Is it possible to have SQL injection here?

Reply:
Changed to a different way. Has it become better?